### PR TITLE
Feat: add role validator

### PR DESCRIPTION
**Pull request by TomerHeber**
Part of #816 

### Solution

Applied more refactoring. In this PR I added a validator and added a test for it.
Made some security fixes in go.mod.

Keeping this ticket open. last part is to deprecate the custom_role_id and role, and use role_id instead (for resourceTeamProjectAssignment).

### CODE REVIEW

File: env0/resource_team_organization_assignment.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ func resourceTeamOrganizationAssignment() *schema.Resource {
 				Type:             schema.TypeString,
 				Description:      "id of the assigned custom role. The following built-in roles can be passed as well: `User`, `Admin`",
 				Required:         true,
-				ValidateDiagFunc: ValidateNotEmptyString,
+				ValidateDiagFunc: NewRoleValidator([]string{"User", "Admin"}),
```
Comment (razbensimon) [inline env0/resource_team_organization_assignment.go:32]: I feel like `[]string{"User", "Admin"}` should be a constant instance that can be imported from `builtin_roles.go`
same for other scopes

File: client/builtin_roles.go [hunk 2]
```diff
@@ -7,6 +7,6 @@ const (
 	Viewer   string = "Viewer"
 )
 
-func IsBuiltinProjectRole(role string) bool {
+func IsBuiltinRole(role string) bool {
```
Comment (razbensimon) [inline client/builtin_roles.go:10]: umm but `IsBuiltinProjectRole` is actually more accurate

File: env0/validators.go [hunk 3]
```diff
@@ -129,3 +117,29 @@ func ValidateTtl(i interface{}, path cty.Path) diag.Diagnostics {
 
 	return nil
 }
+
+func NewRoleValidator(supportedBuiltInRoles []string) schema.SchemaValidateDiagFunc {
```
Comment (razbensimon) [inline env0/validators.go:121]: I don't thing `New` as a prefix in the name is a good choice.
New until when? 🙃

can it be just `RoleValidator`?

File: env0/validators.go [hunk 4]
```diff
@@ -129,3 +117,29 @@ func ValidateTtl(i interface{}, path cty.Path) diag.Diagnostics {
...
+	return func(i interface{}, p cty.Path) diag.Diagnostics {
+		role := i.(string)
+
+		if role == "" {
+			return diag.Errorf("may not be empty")
+		}
+
+		if !client.IsBuiltinRole(role) {
+			// Custom role.
```
Comment (razbensimon) [inline env0/validators.go:130]: even better to create `IsCustomRole` for explicit readability.
inside the common file

File: env0/validators.go (hunk 3 above)
Comment (TomerHeber) [reply env0/validators.go:121]: It creates a call back function that acts as the validator (based on the given input).
That's why a validator is used.

File: env0/resource_team_organization_assignment.go (hunk 1 above)
Comment (TomerHeber) [reply env0/resource_team_organization_assignment.go:32]: Sure.

File: env0/validators.go (hunk 4 above)
Comment (TomerHeber) [reply env0/validators.go:130]: Sure.

**Comment by TomerHeber** [comment]
@razbensimon see commit for additional fixes:
https://github.com/env0/terraform-provider-env0/pull/832/commits/6ed9a03d683f8af5be2fd031448322b5d925f318

**Review by razbensimon — APPROVED** [review]
Approving but still think the comment about the "new" word in the name should be addressed.

File: env0/validators.go (hunk 3 above)
Comment (razbensimon) [reply env0/validators.go:121]: I didnt get your response, I am talking about the function name

Comment (TomerHeber) [reply env0/validators.go:121]: Notice it returns a new function.
In other words, it creates something.
Some validators do not create anything and are without the "new" some do and have the "new".

Comment (TomerHeber) [reply env0/validators.go:121]: I'll merge - if it's still an issue.
We can discuss more - the issue will remain open anyways, as I plan another final PR.

**Pull request merged**